File: website/docs/memory.ts
import { code, note } from "../markup";
import type { Page } from "../types";

export const memory: Page = {
  path: "docs/memory/index.html",
  title: "Memory and recall",
  description: "Know what becomes durable, how search works, how to correct stored context, and what each response budget changes.",
  kind: "docs",
  toc: [
    { label: "Durable memory", id: "durable" },
    { label: "Global and project scope", id: "scope" },
    { label: "Import existing memory", id: "import" },
    { label: "Search and recall", id: "recall" },
    { label: "Inspect and correct", id: "control" },
    { label: "Correcting a memory", id: "supersede" },
    { label: "Response budgets", id: "budgets" },
    { label: "Destructive actions", id: "destructive" },
  ],
  body: `
    <h2 id="durable">What belongs in durable memory</h2>
    <p>Good memory remains useful after the current conversation ends. Store confirmed decisions, recurring preferences, corrected assumptions, naming conventions, project constraints, and facts whose source is clear.</p>
    <p>Shared guidance lives in <code>SOUL.md</code>. Each connected tool's global instruction file contains a managed pointer to it, and the MCP server loads the same file during initialization. Both tools therefore receive the same recall, storage, scope, and token-budget policy.</p>
    <p>Do not store transient task status, speculative ideas, full conversation transcripts, secret values, or instructions that attempt to override the current request or repository guidance. Synapse is the canonical durable memory store; it does not create separate Markdown files for individual memories.</p>
    ${code("shell", `printf '%s\n' 'Use Bun for JavaScript tasks in this repository.' \\
  | synapse memory add tooling`)}
    <p><code>memory add</code> and <code>memory edit</code> read the body from stdin. This keeps long content and shell metacharacters out of command arguments. Empty or whitespace-only memory is rejected.</p>

    <h2 id="scope">Global and project scope</h2>
    <p>Project memory is the default. Synapse resolves the supplied path to its repository or approved project root and returns that memory only alongside global memory for the same project. Unrelated project memory stays out of MCP recall.</p>
    <p>Use global scope only for preferences and conventions that should follow you everywhere. The desktop editor can move a record between scopes. The CLI accepts an explicit scope when adding:</p>
${code("shell", `printf '%s\n' 'Use Bun in this repository.' | synapse memory add tooling --project .
printf '%s\n' 'Prefer concise status updates.' | synapse memory add preferences --global`)}

    <h2 id="import">Import existing memory</h2>
    <p>The Memory screen previews Claude and Codex independently, maps each item to its project when possible, and imports safe entries into the same scoped store. Source files are read-only: Synapse never edits or deletes them.</p>
    <ul>
      <li>Claude imports leaf Markdown files from project memory directories and skips <code>MEMORY.md</code> indexes.</li>
      <li>Codex imports processed memory rows from a recognized local memory database and maps thread IDs to project roots through the local thread catalog.</li>
      <li>Full conversations, session history, tasks, settings, authentication files, and global instruction files are not imported.</li>
      <li>Credential-shaped entries are held for review. Their content is hidden in previews and is not imported by the app.</li>
      <li>Repeated imports are idempotent. Exact scoped duplicates are linked instead of copied, and each batch can be undone without removing a memory that was edited or shared by another source.</li>
    </ul>
${code("shell", `synapse memory import claude
synapse memory import claude --confirm
synapse memory import codex --confirm
synapse memory import markdown ./notes --confirm
synapse memory imports
synapse memory undo <batch> --confirm`)}
${note("Review before overriding the guard", "The CLI can include flagged entries only with both --include-flagged and --confirm. Review the named source file first; memory is plain text and available to connected tools.")}

    <h2 id="recall">Search and recall</h2>
    <p>Memory is stored in an SQLite FTS5 table using Unicode tokenization. A non-empty query searches the memory body; an empty query returns recent entries. Connected tools are instructed to begin with a focused query, the smallest useful result limit, and a Lean response.</p>
${code("json", `{
  "query": "project naming conventions",
  "limit": 4,
  "budget": "lean",
  "project": "/Users/example/project"
}`)}
    <p>The optional per-call <code>budget</code> may be <code>full</code>, <code>balanced</code>, or <code>lean</code>. It can only make the response smaller than the mode selected in Settings; a tool cannot override a user-selected Lean ceiling with Full.</p>
    ${code("shell", `synapse memory list "Bun JavaScript"
synapse memory list --json
synapse memory show 24
synapse memory show 24 --json`)}
    <p>The CLI lists up to 100 matching memories across the local store for management. Each record contains an integer ID, exact body, source, global or project scope, project root, and Unix timestamp. JSON output is suitable for local automation.</p>
    <p>Ranked search drops words that appear in nearly every memory \u2014 <em>the</em>, <em>are</em>, <em>where</em> \u2014 because a memory that matches only on one of those is a confident wrong answer, and an agent acts on it. Words that carry meaning in a preference, including <em>not</em>, <em>never</em>, and <em>use</em>, are kept. Two commands exist for when that is the wrong behaviour or the wrong result:</p>
    ${code("shell", `synapse memory grep -- --no-verify
synapse memory list where are the credentials --explain`)}
    <p><code>memory grep</code> matches the characters you give it and nothing else, which is what you want for an identifier, a flag, a path, or a word the ranked search treats as noise. <code>--explain</code> answers the other question \u2014 why a memory you know is stored did not come back:</p>
    ${code("text", `Query:      where are the credentials
Mode:       search
Expression: "credentials"
Searched:   credentials
Dropped:    where, are, the (matches nearly every memory)
Matches:    1

-0.4596	31	global	vault	Credentials live in Keychain, never in the repository.`)}

    <h2 id="control">Inspect and correct</h2>
    <p>The desktop <strong>Memories</strong> screen searches the same store and lets you inspect, edit, or delete individual entries. CLI edits replace the body and optionally the source:</p>
    ${code("shell", `printf '%s\n' 'Use Bun unless a task explicitly requires another runtime.' \\
  | synapse memory edit 24 tooling

synapse memory show 24`)}
    <p>Editing changes the stored original. Recall optimization does not.</p>

    <h2 id="supersede">Correcting a memory</h2>
    <p>A convention changes. Something you stored last month is now wrong. Adding the new version on its own leaves two memories contradicting each other, both are recalled, and the ranking decides which one a tool acts on \u2014 possibly the one you had already retracted.</p>
    <p>Say which replaced which:</p>
    ${code("shell", `printf '%s\n' 'Deploys run from the release branch, after the tag is signed.' \\
  | synapse memory add deploys --global

synapse memory supersede 12 47`)}
    <p>A connected tool does it in one call, passing the id that came back from its own <code>recall</code>:</p>
    ${code("json", `{
  "content": "Deploys run from the release branch, after the tag is signed.",
  "scope": "global",
  "supersedes": [12]
}`)}
    <p>Nothing is deleted. Memory 12 keeps its id, stays in <code>memory list</code> and the dashboards marked as replaced, still says what replaced it, and comes back at any time:</p>
    ${code("shell", `synapse memory show 12
synapse memory restore 12`)}
    <p>What changes is only what recall can see. A superseded memory stops being returned to tools and stops counting toward the number a session reports \u2014 the count a tool announces is the count it can actually draw on. Deleting the replacement restores what it replaced on its own, so a correction can never leave the older version hidden behind an id that no longer exists.</p>
    <p>Choose between the three: <strong>edit</strong> when the memory was badly worded and there is nothing to keep, <strong>supersede</strong> when the old version was true and stopped being true, and <strong>delete</strong> when it should never have been stored.</p>
    ${note("Supersession stays on this machine", "Sync carries a memory's content, not what replaced it. A memory superseded here arrives on another machine live. Until the wire format grows an operation for it, correct it on each machine you sync to.")}

    <h2 id="budgets">Response budgets</h2>
    <table>
      <thead><tr><th>Mode</th><th>Result limit</th><th>Character budget</th><th>Transformation</th></tr></thead>
      <tbody>
        <tr><td>Full</td><td>25</td><td>Unlimited</td><td>Returns stored formatting without compaction.</td></tr>
        <tr><td>Balanced</td><td>8</td><td>6,000</td><td>Compacts prose whitespace, preserves fenced and indented code, removes exact duplicate bodies, and replaces any memory too large for the remaining budget with its opening sentence.</td></tr>
        <tr><td>Lean</td><td>4</td><td>2,800</td><td>Uses the same non-destructive compaction with a smaller response.</td></tr>
      </tbody>
    </table>
    <p>The response reports the optimization mode actually applied. Choose the ceiling in <strong>Settings → Recall optimization</strong>, or use the CLI:</p>
    ${code("shell", `synapse settings show
synapse settings optimize lean
synapse settings optimize balanced`)}
    <p>A memory that will not fit in what is left of the budget is returned as its opening sentence, marked <code>abridged</code>, rather than cut off wherever the character count landed. Half a memory reads exactly like a whole one \u2014 <em>never deploy from main unless</em> is a rule with its condition amputated \u2014 and one long memory at the top of the results no longer costs you every result under it. A tool that receives an abridged memory is told to recall it again, more narrowly, before acting on the part it cannot see.</p>
    ${note("Originals remain intact", "Balanced and Lean transform only the MCP response. The database and memory-management views keep the complete stored body, abridged results included.")}

    <h2 id="destructive">Destructive actions</h2>
    <p>Deleting one memory, undoing an import, and wiping every memory require explicit confirmation. The desktop app presents a separate confirmation. A wipe removes memory and import history but leaves guidance, vault labels, Keychain values, scope approvals, and settings intact.</p>
    ${code("shell", `synapse memory delete 24 --confirm
synapse memory wipe --confirm`)}
    <p>Before a large cleanup, create a portable snapshot with <code>synapse data export</code>. Use a wipe when the goal is to clear context while retaining the vault setup; use a database restore when the goal is to return the entire Synapse state to an earlier snapshot.</p>
  `,
};

File: crates/synapsecore/src/cli/memory.rs
use crate::brain::{Brain, MemoryScope};
use crate::cli::Outcome;
use crate::imports::{ImportPreview, ImportProvider, ImportStatus};
use anyhow::{Context, Result};
use std::ffi::OsString;
use std::io::{IsTerminal, Read};
use std::path::PathBuf;

const USAGE: &str = "usage: synapse memory <list|grep|show|add|edit|supersede|restore|import|imports|undo|delete|wipe>";

pub fn run(arguments: &[OsString]) -> Result<Outcome> {
    let action = textarg(arguments, 0, USAGE)?;
    let runtime = tokio::runtime::Runtime::new()?;
    let brain = runtime.block_on(Brain::open(crate::files::database()?))?;
    let json = has(before(arguments), "--json");
    match action.as_str() {
        "list" => list(&runtime, &brain, arguments, json)?,
        "grep" => grep(&runtime, &brain, arguments, json)?,
        "show" => show(&runtime, &brain, arguments, json)?,
        "add" => add(&runtime, &brain, arguments)?,
        "edit" => edit(&runtime, &brain, arguments)?,
        "supersede" => supersede(&runtime, &brain, arguments)?,
        "restore" => restore(&runtime, &brain, arguments)?,
        "import" => import(&runtime, &brain, arguments, json)?,
        "imports" => imports(&runtime, &brain, json)?,
        "undo" => undo(&runtime, &brain, arguments)?,
        "delete" => delete(&runtime, &brain, arguments)?,
        "wipe" => wipe(&runtime, &brain, arguments)?,
        _ => anyhow::bail!("unknown memory command `{action}`"),
    }
    Ok(Outcome::Exit(0))
}

fn list(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    arguments: &[OsString],
    json: bool,
) -> Result<()> {
    let query = words(arguments, &["--json", "--explain"]);
    if has(arguments, "--explain") {
        return explain(runtime, brain, &query, json);
    }
    let memories = runtime.block_on(brain.search(&query, 100))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&memories)?);
        return Ok(());
    }
    for memory in memories {
        println!("{}", row(&memory));
    }
    Ok(())
}

fn grep(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    arguments: &[OsString],
    json: bool,
) -> Result<()> {
    let pattern = words(arguments, &["--json"]);
    anyhow::ensure!(
        !pattern.trim().is_empty(),
        "usage: synapse memory grep <text> [--json]"
    );
    let memories = runtime.block_on(brain.grep(&pattern, 100))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&memories)?);
        return Ok(());
    }
    for memory in memories {
        println!("{}", row(&memory));
    }
    Ok(())
}

/// What the search actually did, for the case where the answer is wrong and
/// nothing on screen says why.
fn explain(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    query: &str,
    json: bool,
) -> Result<()> {
    let explanation = runtime.block_on(brain.explain(query, 100))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&explanation)?);
        return Ok(());
    }
    println!("Query:      {}", explanation.query);
    println!("Mode:       {}", explanation.mode());
    match &explanation.expression {
        Some(expression) => println!("Expression: {expression}"),
        None => println!(
            "Expression: none — no term left to search for, so the newest memories answered instead"
        ),
    }
    if !explanation.kept.is_empty() {
        println!("Searched:   {}", explanation.kept.join(", "));
    }
    if !explanation.dropped.is_empty() {
        println!(
            "Dropped:    {} (matches nearly every memory)",
            explanation.dropped.join(", ")
        );
    }
    println!("Matches:    {}", explanation.matches.len());
    println!();
    for hit in &explanation.matches {
        let score = hit
            .rank
            .map(|rank| format!("{rank:.4}"))
            .unwrap_or_else(|| "-".to_owned());
        println!("{score}\t{}", row(&hit.memory));
    }
    Ok(())
}

fn supersede(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    arguments: &[OsString],
) -> Result<()> {
    const USAGE: &str = "usage: synapse memory supersede <old> <new>";
    let old = memoryid(arguments, 1, USAGE)?;
    let new = memoryid(arguments, 2, USAGE)?;
    anyhow::ensure!(
        runtime.block_on(brain.supersede(old, new))?,
        "memory #{old} not found"
    );
    println!("Memory #{old} superseded by #{new}; recall now returns #{new} instead");
    println!("Undo with: synapse memory restore {old}");
    Ok(())
}

fn restore(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    let id = memoryid(arguments, 1, "usage: synapse memory restore <id>")?;
    anyhow::ensure!(
        runtime.block_on(brain.unsupersede(id))?,
        "memory #{id} is not superseded"
    );
    println!("Restored memory #{id}");
    Ok(())
}

/// One memory as a line: id, scope, source, and enough body to recognise it.
fn row(memory: &crate::brain::Memory) -> String {
    let scope = if memory.scope == MemoryScope::Global {
        "global".to_owned()
    } else {
        format!("project:{}", memory.project)
    };
    let body = if memory.superseded == 0 {
        bodypreview(&memory.body)
    } else {
        format!(
            "(superseded by #{}) {}",
            memory.superseded,
            bodypreview(&memory.body)
        )
    };
    format!(
        "{}\t{}\t{}\t{}",
        memory.id,
        scope,
        if memory.source.is_empty() {
            "-"
        } else {
            &memory.source
        },
        body
    )
}

/// The positional words of a subcommand, with the named flags taken out.
///
/// A bare `--` ends the flags: everything after it is the text being searched
/// for, which is what lets `memory grep -- --no-verify` mean the flag rather
/// than a flag.
fn words(arguments: &[OsString], flags: &[&str]) -> String {
    let rest = arguments.get(1..).unwrap_or_default();
    if let Some(at) = rest.iter().position(|value| value == "--") {
        return join(rest[at + 1..].iter());
    }
    join(
        rest.iter()
            .filter(|value| !flags.iter().any(|flag| value == flag)),
    )
}

fn join<'a>(values: impl Iterator<Item = &'a OsString>) -> String {
    values
        .map(|value| value.to_string_lossy())
        .collect::<Vec<_>>()
        .join(" ")
}

/// Arguments up to a bare `--`. A flag written after the separator is part of
/// the text being searched for, not a flag.
fn before(arguments: &[OsString]) -> &[OsString] {
    match arguments.iter().position(|value| value == "--") {
        Some(at) => &arguments[..at],
        None => arguments,
    }
}

fn show(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    arguments: &[OsString],
    json: bool,
) -> Result<()> {
    let id = memoryid(arguments, 1, "usage: synapse memory show <id> [--json]")?;
    let memory = runtime
        .block_on(brain.memory(id))?
        .context("memory not found")?;
    if json {
        println!("{}", serde_json::to_string_pretty(&memory)?);
    } else {
        println!("Memory #{}", memory.id);
        println!("Scope: {}", memory.scope.value());
        if !memory.project.is_empty() {
            println!("Project: {}", memory.project);
        }
        println!("Source: {}", memory.source);
        println!("Created: {}", memory.created);
        if memory.superseded != 0 {
            println!(
                "Superseded by: #{} (not recalled; restore with `synapse memory restore {}`)",
                memory.superseded, memory.id
            );
        }
        println!();
        println!("{}", memory.body);
    }
    Ok(())
}

fn add(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    let (scope, project) = memoryscope(arguments)?;
    let source = positional(arguments, 1, &["--global", "--project"]);
    let body = readmemory()?;
    let id = runtime.block_on(brain.rememberscoped(
        &body,
        source.as_deref(),
        scope,
        project.as_deref(),
    ))?;
    println!("Stored memory #{id}");
    Ok(())
}

fn edit(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    let id = memoryid(arguments, 1, "usage: synapse memory edit <id> [source]")?;
    let source = arguments.get(2).and_then(|value| value.to_str());
    let body = readmemory()?;
    runtime
        .block_on(brain.updatememory(id, &body, source))?
        .context("memory not found")?;
    println!("Updated memory #{id}");
    Ok(())
}

fn import(
    runtime: &tokio::runtime::Runtime,
    brain: &Brain,
    arguments: &[OsString],
    json: bool,
) -> Result<()> {
    let provider: ImportProvider = textarg(
        arguments,
        1,
        "usage: synapse memory import <claude|codex|markdown> [path] [--confirm]",
    )?
    .parse()?;
    let candidates = if provider == ImportProvider::Markdown {
        let path = PathBuf::from(textarg(
            arguments,
            2,
            "usage: synapse memory import markdown <path> [--confirm]",
        )?);
        crate::imports::scanmarkdown(&path)?
    } else {
        let home = crate::files::home()?;
        runtime.block_on(crate::imports::scan(&home, provider))?
    };
    let preview = runtime.block_on(brain.importpreview(provider, candidates))?;
    if !has(arguments, "--confirm") {
        printpreview(&preview, json)?;
        if !json {
            println!("Preview only. Add --confirm to import safe entries.");
        }
        return Ok(());
    }
    let includeflagged = has(arguments, "--include-flagged");
    let report = runtime.block_on(brain.importmemories(preview, includeflagged))?;
    if json {
        println!("{}", serde_json::to_string_pretty(&report)?);
    } else {
        println!(
            "Import batch #{}: {} stored, {} linked, {} already imported, {} flagged and left untouched",
            report.batch.id, report.stored, report.linked, report.skipped, report.flagged
        );
        println!(
            "Undo with: synapse memory undo {} --confirm",
            report.batch.id
        );
    }
    Ok(())
}

fn imports(runtime: &tokio::runtime::Runtime, brain: &Brain, json: bool) -> Result<()> {
    let batches = runtime.block_on(brain.importbatches())?;
    if json {
        println!("{}", serde_json::to_string_pretty(&batches)?);
        return Ok(());
    }
    for batch in batches {
        println!(
            "{}\t{}\t{} stored\t{} linked\t{}",
            batch.id,
            batch.provider,
            batch.imported,
            batch.linked,
            if batch.undone { "undone" } else { "active" }
        );
    }
    Ok(())
}

fn undo(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    let id = memoryid(arguments, 1, "usage: synapse memory undo <batch> --confirm")?;
    anyhow::ensure!(
        has(arguments, "--confirm"),
        "add --confirm to undo import batch {id}"
    );
    let deleted = runtime.block_on(brain.undoimport(id))?;
    println!("Undid import batch #{id}; removed {deleted} imported memories");
    Ok(())
}

fn delete(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    let id = memoryid(arguments, 1, "usage: synapse memory delete <id> --confirm")?;
    anyhow::ensure!(
        has(arguments, "--confirm"),
        "add --confirm to delete memory #{id}"
    );
    runtime
        .block_on(brain.deletememory(id))?
        .context("memory not found")?;
    println!("Deleted memory #{id}");
    Ok(())
}

fn wipe(runtime: &tokio::runtime::Runtime, brain: &Brain, arguments: &[OsString]) -> Result<()> {
    anyhow::ensure!(
        has(arguments, "--confirm"),
        "add --confirm to delete every memory"
    );
    let count = runtime.block_on(brain.wipememories())?;
    println!("Deleted {count} memories");
    Ok(())
}

fn memoryscope(arguments: &[OsString]) -> Result<(MemoryScope, Option<PathBuf>)> {
    anyhow::ensure!(
        !(has(arguments, "--global") && has(arguments, "--project")),
        "choose either --global or --project"
    );
    if has(arguments, "--global") {
        return Ok((MemoryScope::Global, None));
    }
    let project = optionvalue(arguments, "--project")
        .map(PathBuf::from)
        .unwrap_or(std::env::current_dir()?);
    Ok((MemoryScope::Project, Some(project)))
}

fn printpreview(preview: &ImportPreview, json: bool) -> Result<()> {
    if json {
        println!("{}", serde_json::to_string_pretty(preview)?);
        return Ok(());
    }
    println!(
        "{}: {} found · {} ready · {} existing · {} flagged",
        preview.provider.name(),
        preview.found,
        preview.ready,
        preview.existing,
        preview.flagged
    );
    for item in &preview.items {
        let marker = match item.status {
            ImportStatus::Ready => "ready",
            ImportStatus::Existing => "existing",
            ImportStatus::Flagged => "review",
        };
        println!("{marker}\t{}\t{}", item.path, item.preview);
        if let Some(warning) = &item.warning {
            println!("\twarning: {warning}");
        }
    }
    Ok(())
}

fn readmemory() -> Result<String> {
    anyhow::ensure!(
        !std::io::stdin().is_terminal(),
        "pipe the replacement memory content to stdin"
    );
    let mut value = String::new();
    std::io::stdin().read_to_string(&mut value)?;
    anyhow::ensure!(!value.trim().is_empty(), "memory content cannot be empty");
    Ok(value)
}

fn positional(arguments: &[OsString], start: usize, options: &[&str]) -> Option<String> {
    let mut skip = false;
    for value in arguments.iter().skip(start) {
        if skip {
            skip = false;
            continue;
        }
        let value = value.to_str()?;
        if value == "--project" {
            skip = true;
            continue;
        }
        if options.contains(&value) || value.starts_with("--") {
            continue;
        }
        return Some(value.to_owned());
    }
    None
}

fn optionvalue(arguments: &[OsString], option: &str) -> Option<String> {
    arguments
        .iter()
        .position(|value| value == option)
        .and_then(|index| arguments.get(index + 1))
        .and_then(|value| value.to_str())
        .map(ToOwned::to_owned)
}

fn memoryid(arguments: &[OsString], index: usize, usage: &str) -> Result<i64> {
    textarg(arguments, index, usage)?
        .parse()
        .with_context(|| usage.to_owned())
}

fn bodypreview(value: &str) -> String {
    let compact = value.split_whitespace().collect::<Vec<_>>().join(" ");
    let mut characters = compact.chars();
    let preview = characters.by_ref().take(80).collect::<String>();
    if characters.next().is_some() {
        format!("{preview}…")
    } else {
        preview
    }
}

fn has(arguments: &[OsString], flag: &str) -> bool {
    arguments.iter().any(|value| value == flag)
}

fn textarg(arguments: &[OsString], index: usize, usage: &str) -> Result<String> {
    arguments
        .get(index)
        .and_then(|value| value.to_str())
        .map(ToOwned::to_owned)
        .with_context(|| usage.to_owned())
}
